### ISSUE TITLE
:! (bang) and system() are not interactive / show ANSI codes

### ISSUE THREAD

**Issue by Tranquility** [neovim/neovim#1496]
If I type `!git add -p` it loops over the output and only stops if i press ctrl-c.

**Comment by elmart** [comment]
Will #2076 fix this?

**Comment by tarruda** [comment]
@elmart no but it will offer an alternative: `:term git add -p`. The downside is that plugins relying on spawning interactive programs will have to be adapted for neovim. Fortunately there doesn't seem to be many such plugins out there, probably a consequence of the unreliability when running gvim.

At first I considered reimplementing bang commands on top of the feature introduced in #2076 but came to conclusion that would be compromising too much just for compatiblity. See [this](https://github.com/junegunn/fzf/issues/137#issuecomment-77713710) for an explanation.

**Comment by justinmk** [comment]
> no but it will offer an alternative: `:term git add -p`.

vim-fugitive did so in https://github.com/tpope/vim-fugitive/pull/624.

**Comment by ggVGc** [comment]
What is the suggested workaround for launching an interactive :! command in cmdline mode?

**Comment by justinmk** [comment]
Use `:te` instead.

**Comment by ggVGc** [comment]
As far as I understand, :terminal is not allowed from cmdline mode, which is why I asked.
``` vim
fun! Test()
  :term echo "Test!"
endfun
cnoremap <silent> <c-n> <c-\>eTest()<cr>
```
gives me `E523: Not allowed here: :enew | call termopen("echo \"Test!\"") | startinsert`

**Comment by justinmk** [comment]
Looks like a bug. We may need to shellescape?

**Comment by ZyX-I** [comment]
@justinmk Because this is not true command mode. `<C-\>e` is used to edit the command-line and allowing `:term` (or any command that does something with buffers) is not a good idea. I would not be surprised to hear that it has technical reason behind this. My usual workaround for such kind of things is messing with `feedkeys()`.

**Comment by justinmk** [comment]
Reopening this to track this new use-case.

**Comment by ggVGc** [comment]
Fwiw I think it's very useful to be able to be able to call external programs and use their output in `cmdline`. I agree commands dealing with buffers should probably be disallowed in cmdline mode, but the issue here is that the behaviour of :! has been changed with regards to vim, and :term is the replacement for interactive commands. So in this case we've actually lost integration with the host system rather than gained some, which as far as I understand is one of the main project goals of neovim.

**Comment by justinmk** [comment]
@ggVGc By "new use case" I mean "new information that was not previously written down".

**Comment by ZyX-I** [comment]
@ggVGc `system()` should still be working. `!` has nothing to do with “using output in cmdline”. I have actually no idea why would `!` be useful in `<C-\>e` expression or `<expr>` mapping: I would never imagine that calling *any* interactive program *in this context* has any sense except for making me angry and for other programs `system()` should be fine. (Though mostly because this is hard context switching (altscreen gets replaced with regular screen), if I need to e.g. output commit hash to the command-line I have nothing againts using `input()` in a mapping.)

**Comment by justinmk** [comment]
@ggVGc Does `input()` and `system()` cover your use-case? (Of course `system()` is also not interactive)

**Comment by tarruda** [comment]
> So in this case we've actually lost integration with the host system rather than gained some, which as far as I understand is one of the main project goals of neovim.

`termopen()` lets you integrate nvim with interactive programs, have you seen fzf? While we diverged the behavior of `system()` and `!`, it was for allowing the full client server architecture planned from the beginning of the project. We can't have UIs running in different processes and have programs require nvim stdio inheritance in order to display output.

Also, its not like vim behavior is consistent anyway, gvim also has limited support for interactive programs in `!` and `system()`(which is why plugins like fzf and selecta dont work in gvim).

I hope we'll eventually be able to run full interactive programs with `!` and `system()`, this should be possible by making some enhancements to the terminal module and have it draw directly to the screen(instead of a buffer), but it is not a primary goal.

**Comment by ggVGc** [comment]
Yeah, I get it, and I'm not arguing the tradeoffs. Just pointing out that something that's previously worked(and I think has been very useful), for now doesn't in nvim. But I understand the community already knows about this. Fzf is actually what my use case was, I have been using fzf to bring up a fuzzy list of completions in `cmdline` mode. So, i.e :e <c-n> would give me a fuzzy list of all files in the current cwd to open, or :h f<c-n> gives me a list of all help entries starting with f.

**Comment by acornejo** [comment]
there are some open issues in terminal that affect fzf. for example, junegunn/fzf#426

**Comment by amerlyq** [comment]
If there is no support for interactive commands, then how about some ways to control main terminal?
I understand, we currently have env variables to enable cursor shapes and already integrated Focus* events not long ago. But for ```urxvt```, which has no focus support, I need enable/disable focus events manually, by sending escape sequence ```"\e]777;focus;on\x7"``` to corresponding custom urxvt extension ([vim-focus-autocmd](https://github.com/amerlyq/vim-focus-autocmd) + [urxvt-ext-evolved](https://github.com/amerlyq/urxvt-ext-evolved)). 

Therefore, as ```t_ti/t_te``` don't work now, I could use ```silent !echo -e "\e]777;focus;on\x7"``` instead.
But it don't work anymore. How could I send this escape sequence to main terminal?

**Comment by justinmk** [comment]
It is planned.

**Comment by Asher256** [comment]
Why not ? Some outside programs could need this feature.

I would be interesting if you have implemented the previous vim behavior of :shell with :lshell (legacy shell command, for example).

In my case, I prefer to start programs outside Neovim (especially because currently, :terminal is sometimes buggy. Especially when I run Python 3 programs. This bug doesn't happens every time and I don't know whats causes this).

**Comment by justinmk** [comment]
No, that is not going to happen. If there is a bug with :term it should be fixed. Any partial restoration of :shell will use :term internally. 

Please don't report things without reduced steps to reproduce with a specific version of nvim.

**Comment by sassanh** [comment]
How I wish this part of Thiago's comment was real: "I'm starting to think it's better to find a solution to that soon."

**Comment by phodge** [comment]
As a tmux users may find this mapping useful:
```    if has('nvim')
      cnoremap <expr> ! strlen(getcmdline())?'!':('!tmux split-window -c '.getcwd().' ')
    endif
```

**Comment by KillTheMule** [comment]
He, nice one :D

**Comment by goerz** [comment]
This is a terrible design decision! There should be a way to run a command in the full underlying shell without quitting or suspending nvim. This breaks just so many use cases, as all the bug reports referencing this issue show. Personally, I've always done `!less %` for getting a quick view of the current file for copy/paste in the terminal without disrupting my vim layout. This now is no longer possible (I'd now have do `Ctrl-z`, `less <filename>`, `fg`). If you're going to change the behavior of `!`, at least provide some different mechanism to achieve the old behavior, so I can create a custom mapping!

**Comment by ZyX-I** [comment]
@goerz There is `:terminal` for the task currently, as mentioned some comments above and in some other issues. There is some desire to implement the equivalent for Vim’s `guioptions+=!`, but this time Bram managed to beat us and I am not sure whether somebody is working on that. @justinmk should know better.

**Comment by goerz** [comment]
If https://github.com/neovim/neovim/issues/5541 was implemented, that would also allow for a workaround for the problem of opening an interactive process using the full width of the terminal, without disrupting any window layouts in vim. I would be able to do

    :tab te less %

**Comment by janlazo** [comment]
https://github.com/janlazo/dotvim8/blob/master/autoload/dotvim8.vim#L79-L124 is my backward-compatible workaround for `:!` via `:terminal` which I use in a [batchfile](https://github.com/janlazo/dotvim8/blob/master/after/ftplugin/dosbatch.vim) or [powershell script](https://github.com/janlazo/dotvim8/blob/master/after/ftplugin/ps1.vim) for `K`.

It's a shim but I'm satisfied with it and I don't need `guioptions+=!`

**Comment by janlazo** [comment]
I started https://github.com/janlazo/dotvim8/commit/d0f9026a3a8ac7a259964c710acb4618c8ff7b66 to swap `:!` with my workaround on GVim. Feedback is appreciated.

**Comment by janlazo** [comment]
I extracted my workaround and put it in a separate repo, https://github.com/janlazo/vim-bang-terminal.
Feedback is appreciated.

**Comment by joshuaunderwood7** [comment]
Yeah, this is a deal breaker.
As a new user who was excited about nvim, I'm forced into dropping it until this is dealt with.  NeoVim is not a vim replacement without implementing the current features; and this is not esoteric. This is pretty fundamental vim functionality, and pretty common among programs that implement vim-like interfaces.
Rather than close all the issues, this should be considered a critical bug and resolved (now 5 years after the original complaint and still going strong).

**Comment by bfredl** [comment]
> Rather than close all the issues, this should be considered a critical bug and resolved (now 5 years after the original complaint and still going strong).

But this issue is open. Only duplicate issues are closed.

> Rather than close all the issues, this should be considered a critical bug and resolved (now 5 years after the original complaint and still going strong).

It is planned to be fixed in 0.5, see #10842 . Note that gvim did not support this until 1-2 years ago, I don't think one would consider gvim to be lacking "fundamental vim functionality" until then.

**Comment by sassanh** [comment]
@josephholsten I'm using Neovim for more than 3 years, I spend something like 6 hours each day in Neovim and trust me it's not a big/medium/small deal at all, nothing relies on this, I have more than 150 plugins and I never noticed this issue past 3 years. I'm subscribed to this issue because in the beginning it was super weird for me that this issue exist on Neovim and I thought I can't use Neovim until it's solved, so I subscribed to this issue but in next few hours I found out how to update my vimrc (`init.vim`) and then I forgot this issue even exist. I don't remember the changes I had to do, but I remember it didn't take more than few hours.

Please feel free to list the things that are affected by this issue and are annoying you and we'll try to provide you solutions.

**Comment by josephholsten** [comment]
@sassanh I suspect you meant someone else ;-)

**Comment by joshuaunderwood7** [comment]
> > Rather than close all the issues, this should be considered a critical bug and resolved (now 5 years after the original complaint and still going strong).
> 
> But this issue is open. Only duplicate issues are closed.
> 
> > Rather than close all the issues, this should be considered a critical bug and resolved (now 5 years after the original complaint and still going strong).
> 
> It is planned to be fixed in 0.5, see #10842 . Note that gvim did not support this until 1-2 years ago, I don't think one would consider gvim to be lacking "fundamental vim functionality" until then.

Arrogant fucking programmers...

Ok, what-the-fuck ever. Do as you will. I’ll keep using real vim.

**Comment by justinhoward** [comment]
While I don't sympathize with Joshua's conclusion, this problem did surprise me when I first encountered it. It's by no means a deal-breaker, but I wouldn't say it's a non-issue either. Particularly, I miss `:w !sudo tee %` mentioned in #1716, but I installed the suda.vim plugin and moved on.

**Comment by justinmk** [comment]
Trust me, many of us want this fixed. We spent our time-units on other things. Someone needs to spend life-minutes fixing this, then it will be fixed.

**Comment by zeertzjq** [comment]
"msgsep" has been graduated in #20374, so a possible approach to this is to use a terminal floating window positioned in the message area that draws directly on the message grid.
